### PR TITLE
Fix attention mask creation for GPTNeo

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

It seems that #26486 broke the way the attention mask was created. It creates a causal attention mask by default, but there is already a causal attention mask in the `GPTNeoSelfAttention` modules, resulting in `NaN`s.

I am not sure the solution is perfect, so opened to suggestions.

### CODE REVIEW

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 1]
```diff
@@ -802,10 +802,29 @@ def forward(
...
+            # used in OpenAI GPT, we just need to prepare the broadcast dimension here.
+            attention_mask = attention_mask[:, None, None, :]
+
+            # Since attention_mask is 1.0 for positions we want to attend and 0.0 for
+            # masked positions, this operation will create a tensor which is 0.0 for
+            # positions we want to attend and the dtype's smallest value for masked positions.
+            # Since we are adding it to the raw scores before the softmax, this is
+            # effectively the same as removing these entirely.
+            attention_mask = attention_mask.to(dtype=self.dtype)  # fp16 compatibility
+            attention_mask = (1.0 - attention_mask) * torch.finfo(self.dtype).min
```
Comment (fxmarty) [inline src/transformers/models/gpt_neo/modeling_gpt_neo.py:827]: If you'd like you can use `modeling_attn_mask_utils._prepare_4d_attention_mask` as well

Comment (michaelbenayoun) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py]: Thanks for the suggestion!

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing! 

This should have been caught by tests 😬 Could you add a test (or tests) which would have caught this?

**Review by ArthurZucker — COMMENTED** [review]
+1 on adding a test! 
Also see #27931 for how we'll handle mask in the futur I hope

**Comment by michaelbenayoun** [comment]
I will work on that when I have the bandwidth!

**Comment by amyeroberts** [comment]
Any update @michaelbenayoun?

**Comment by michaelbenayoun** [comment]
I can try to work on that next week.

**Comment by michaelbenayoun** [comment]
While trying to write an example, I noticed that I think it works now...

**Comment by amyeroberts** [comment]
OK - that's good news really! Shall we close the PR?

**Comment by michaelbenayoun** [comment]
Yes closing it.
